Implement windowsDiff.Compare to allow outputting OCI images

TBBle opened the pull request:
This builds on #4395 to implement the `Compare` side of the differ and similarly to `Apply`, adds a redirect for WCOW to `archive.WriteDiff`.

With this change, (and a bunch of buildkit hacks), I can now export an image I built with buildkit to any of the types `oci`, `docker`, or `image`.

Fixes: #4394

TBBle commented:
I've speculatively enabled the snapshot tests on Windows, in case I've fixed them as part of this. If that's not the case, I'll back that enablement out for later.

TBBle commented:
There's a bunch of refactoring opportunities I saw from this work.

Particularly, it seems that the APIs for `archive.WriteDiff` and `archive.Apply` are somewhat nonsensical when applied to windows layers, as we use the `Opts` to smuggle the real parameters through them, and ignore the code in favour of complete Windows-specific implementations. I think these two methods might need more insight into the thing they're being called on, to actually do their job without so much hand-holding.

Apart from tests, `WriteDiff` only appears to be used from the `walkingDiff`, and `windowsDiff` implementations of `Compare`, and for taking task checkpoints. The latter is always diffing against "", and does not interact with Windows Container layers.

`Apply` is used more-widely, but again only `windowsDiff` knows about Windows Container layers.

I suspect it would be cleaner to not involve `archive` in the Windows Container layers, but to directly implement those functions in `windowsDiff`. Particularly if we lean more on hcsshim and go-winio for the actual implementations.

jterry75 commented:
@kevpar / @jstarks - PTAL

jstarks commented:
@ambarve

jterry75 commented:
@ambarve - PTAL

TBBle commented:
Marked as draft pending rebase on updated #4395, which will remove the `winio_tar` aspect of this change.

TBBle commented on archive/tar.go:
```diff
@@ -63,13 +63,34 @@ func Diff(ctx context.Context, a, b string) io.ReadCloser {
+	return options.writeDiffFunc(ctx, w, a, b, options)
+}
+
+// WriteDiff writes a tar stream of the computed difference between the
```
`writeDiffNaive`.

TBBle commented on archive/tar_windows.go:
```diff
@@ -259,3 +259,81 @@ func tarToBackupStreamWithMutatedFiles(buf *bufio.Writer, w io.Writer, t *tar.Re
 	return backuptar.WriteBackupStreamFromTarFile(buf, t, hdr)
 }
+
+// WriteDiff writes a tar stream of the computed difference between the
```
`writeDiffWindowsLayers`

TBBle commented on diff/windows/windows.go:
```diff
@@ -158,7 +163,135 @@ func (s windowsDiff) Apply(ctx context.Context, desc ocispec.Descriptor, mounts
+		}
+	}
+
+	// TODO darrenstahlmsft: When this is done isolated, we should disable this.
+	// it currently cannot be disabled, unless we add ref counting. Since this is
+	// temporary, leaving it enabled is OK for now.
```
Add a reference to #1681, which is _why_ this is "temporary"

TBBle replied:
Maybe fix the comment I copied this from too?

jterry75 commented:
@ambarve / @kevpar - PTAL so we can get these in

TBBle commented:
Once #4395 lands, this can be trivially rebased to get rid of the vendor.conf conflict (which comes from #4395, not this PR's changes)

cpuguy83 commented:
🎉 Looks like the other PR is merged.

TBBle commented:
Yup. I'll try and rebase this soon (ideally one evening this week) and I believe it'll be ready-for-review.

TBBle commented:
That was a trivial rebase, and a much shorter diff to read. This should be ready-to-go now.

jterry75 commented:
@ambarve - Can you please take a look. Especially at the places where ociwclayer.Export was duplicated. I'm concerned that we dont actually want to expose that internal package and we should instead do this via the hcsshim public api. Or we should promote ociwclayer functionality outside of just the wclayer.exe tool. What do you think?

TBBle commented:
The idea of exposing `internal/ociwclayer` from hcsshim as `pkg/ociwclayer` instead of duplicating it here has been favourably received by the hcsshim developers in https://github.com/microsoft/hcsshim/pull/916. So I'm hoping that will land soon (once everyone's back from end-of-solar-year holidays) and then I can rebase this work upon that new build of hcsshim.

I also wouldn't mind if this went in as-is, as I could refactor the duplication away later. There are other (more-essential) PRs in hcsshim and go-winio that have come out of this work, and I'd like to see them merged, and then pull those into containerd to support #4419, the current end-point of this PR chain.

I don't particularly want those to block landing this PR.

jterry75 commented on archive/tar_opts.go:
```diff
@@ -73,3 +73,13 @@ func WithParents(p []string) ApplyOpt {
+
+// WriteDiffOptions provides additional options for a WriteDiff operation
+type WriteDiffOptions struct {
+	ParentLayers []string // Windows needs the full list of parent layers
```
Should this be `WriteDiffOptions` in `tar_opts_windows.go` since its Windows specific?

TBBle replied:
The structure needs to be visible in the cross-platform code as it's passed to `archive.WriteDiff`, unless we want to instead complete decouple WCOW Layer Diff/Apply from using `archive` APIs. (Which I'd be fine with, but is a much wider change, and affects existing APIs)

The particular field is only used on Windows, and I'm not sure it's more valuable than confusing to bifurcate the structure by defining `WriteDiffOptions` differently for `windows` and `!windows`. The same argument could be made for `applyFunc` and `writeDiffFunc`, so I'd rather punt this to later possible refactoring as those are much-worse warts, and there are much greater improvements available than splitting this structure.

An alternative approach to removing this field would be to smuggle this array of strings via the `a` (or `lower`) string parameter which is currently unused in this flow. I think I saw code somewhere around here which was doing that by serialising the array into JSON and the unserialising it once back in platform-specific code.

I was not fond of that...

jterry75 replied:
Np

jterry75 commented on archive/tar_opts_windows.go:
```diff
@@ -27,3 +27,22 @@ func AsWindowsContainerLayer() ApplyOpt {
+
+// AsWindowsContainerLayerPair indicates that the paths to diff are a pair of
+// Windows Container Layers. The caller must be holding SeBackupPrivilege.
+func AsWindowsContainerLayerPair() WriteDiffOpt {
```
I actually dont understand this one? What does this mean to you :)

TBBle replied:
It's like `AsWindowsContainerLayer`, except instead of applying a layer, it's diffing a pair of layers. This whole redirection via `applyFunc`/`writeDiffFunc` is a bit weird, which was my thinking in https://github.com/containerd/containerd/pull/4399#issuecomment-660283335. However, I didn't want to refactor the Apply and Compare paths as part of the same PR that implemented the Compare path to match the Apply path.

katiewasnothere replied:
The place that we actually use this config option, we've already verified the two paths are layers. IMO it makes more sense to make this generic, such as `WithDiffFunc(df diffFunc)` or something like that. What do you think?

TBBle replied:
Hmm. I could refactor `AsWindowsContainerLayer` and `AsWindowsContainerLayerPair` to be `WithApplyFunc` and `WithDiffFunc`, and then `applyWindowsLayer` and `writeDiffWindowsLayer` would move into diff/windows/windows.go, passed in by `(s windowsDiff) Apply` and `(s windowsDiff) Compare` respectively.

However, that means that anyone else trying to use `github.com/containerd/containerd/archive.Apply` or `github.com/containerd/containerd/archive.WriteDiff` would need to provide their own diff funcs to pass in, unless we make `applyWindowsLayer` and `writeDiffWindowsLayer` public, which I'm not super-fond of.

TBBle replied:
Maybe as a clarifying question, how "public" is `github.com/containerd/containerd/archive` support for WCOW layers supposed to be, as an API?

Apart from `github.com/containerd/containerd/diff/windows/`, nothing in this tree seems to be prepared for WCOW usage of `archive.Apply` or `archive.Diff`, so it might make sense to replace `archive.Apply(..., AsWindowsContainerLayer())` and `archive.Diff(..., AsWindowsContainerLayerPair())` with functions directly in `github.com/containerd/containerd/diff/windows/`, and leave `github.com/containerd/containerd/archive` to only support the walking-diff use-case, perhaps with a slight special-case to recognise WCOW layer archives and pretend nothing outside `Files/` exists, i.e. replacing the `strings.Replace(hdr.Name, "Files", "", 1)` in /install.go with one performed during the walk, to allow _some_ rational behaviour for non-Windows platforms manipulating WCOW layer archives.

TBBle replied:
Overall, I'd prefer not to mix such refactorings into this simply because the current approach implements Compare to match Apply, and any further changes should be done to both at once, and as an API change that affects an existing public API, be clearly distinguished as-such.

jterry75 commented on diff/windows/windows.go:
```diff
@@ -158,7 +164,136 @@ func (s windowsDiff) Apply(ctx context.Context, desc ocispec.Descriptor, mounts
+	}
+
+	cw, err := s.store.Writer(ctx, content.WithRef(config.Reference), content.WithDescriptor(ocispec.Descriptor{
+		MediaType: config.MediaType, // most contentstore implementations just ignore this
```
This is an odd comment that may change at any time. Can we just remove it?

TBBle replied:
It was copied over from `func (s *walkingDiff) Compare`. Happy to drop it, if it's not an active concern.

jterry75 commented on diff/windows/windows.go:
```diff
@@ -191,3 +326,52 @@ func mountsToLayerAndParents(mounts []mount.Mount) (string, []string, error) {
+		return nil, errors.Wrap(errdefs.ErrInvalidArgument, "windowsDiff cannot diff a layer with no parents against another layer")
+	}
+
+	// May return an ErrNotImplemented, which will fall back to LCOW
```
Is this still true? SHould we handle the error here?

TBBle replied:
Good point. At this point, that suggests we've been asked to diff a windows-layer against a non-windows-layer. It makes sense to fail properly, as I don't think that's expected to be rational.

TBBle commented:
Marking as "Draft" as I plan to rebase this onto https://github.com/microsoft/hcsshim/pull/916 this weekend. It'll basically be deleting a bunch of code we no longer need to duplicate, and should not affect the structure of this PR, so current reviews remain valid.

Except the concerns about duplicating code out of hcsshim, of course.

TBBle commented:
Current Windows test failure is some flakiness elsewhere; possibly inside HCS, but I currently think it's an API miuse in the test suite.

The failure was
```
2021-01-16T09:52:57.8675958Z signal arrived during external code execution
2021-01-16T09:52:57.8676542Z 
2021-01-16T09:52:57.8683433Z syscall.Syscall(0x7ffda670d430, 0x2, 0x1b40f30, 0xc0195c6780, 0x0, 0x0, 0x0, 0x0)
2021-01-16T09:52:57.8684502Z 	C:/hostedtoolcache/windows/go/1.15.6/x64/src/runtime/syscall_windows.go:188 +0xe9
2021-01-16T09:52:57.8694631Z github.com/Microsoft/hcsshim/internal/wclayer._destroyLayer(0x1b40f30, 0xc0195c6780, 0xc0195c6780, 0x26)
2021-01-16T09:52:57.8696440Z 	D:/a/containerd/containerd/src/github.com/containerd/containerd/vendor/github.com/Microsoft/hcsshim/internal/wclayer/zsyscall_windows.go:236 +0x172
2021-01-16T09:52:57.8698028Z github.com/Microsoft/hcsshim/internal/wclayer.destroyLayer(0x1b40f30, 0xc01959a660, 0x25, 0x1, 0x0)
2021-01-16T09:52:57.8699453Z 	D:/a/containerd/containerd/src/github.com/containerd/containerd/vendor/github.com/Microsoft/hcsshim/internal/wclayer/zsyscall_windows.go:229 +0xbc
2021-01-16T09:52:57.8701005Z github.com/Microsoft/hcsshim/internal/wclayer.DestroyLayer(0x1e4ce40, 0xc0000960d0, 0xc01959a660, 0x25, 0x0, 0x0)
2021-01-16T09:52:57.8702700Z 	D:/a/containerd/containerd/src/github.com/containerd/containerd/vendor/github.com/Microsoft/hcsshim/internal/wclayer/destroylayer.go:20 +0x205
2021-01-16T09:52:57.8704252Z github.com/Microsoft/hcsshim.DestroyLayer(0x0, 0xc0195dc0c0, 0x1b, 0xc00024652c, 0x9, 0x2310ba0, 0xc0195cb220)
2021-01-16T09:52:57.8705604Z 	D:/a/containerd/containerd/src/github.com/containerd/containerd/vendor/github.com/Microsoft/hcsshim/layer.go:34 +0xe5
2021-01-16T09:52:57.8706975Z github.com/containerd/containerd/sys.ForceRemoveAll(0xc000246510, 0x25, 0xc0195cb220, 0x1d26e29)
2021-01-16T09:52:57.8708204Z 	D:/a/containerd/containerd/src/github.com/containerd/containerd/sys/filesys_windows.go:267 +0xaf
2021-01-16T09:52:57.8709334Z github.com/containerd/containerd.TestMain(0xc000135f00)
2021-01-16T09:52:57.8710417Z 	D:/a/containerd/containerd/src/github.com/containerd/containerd/client_test.go:168 +0xe05
2021-01-16T09:52:57.8711228Z main.main()
2021-01-16T09:52:57.8711936Z 	_testmain.go:153 +0x272
```

It's happening during clean-up after the tests via `sys.ForceRemoveAll`. I've been looking into this is #4419 which seems to reproduce it more-often, and I haven't seen this crash since I changed `sys.ForceRemoveAll` to not try and call `DestroyLayer` on random folders. This is the first time I've seen it without the Snapshot testsuite enabled, suggesting it takes a lot of HCS wclayer activity to reproduce, and possibly only happens if the test suite tears down while the layers are still being unloaded.

The [code-identical](https://github.com/containerd/containerd/compare/eb1adede03a990e220fad95cd2614fb8af7edddc..e949362c288aa38e54b1aa9d70fb2cf96b3e47b9) commit https://github.com/containerd/containerd/commit/eb1adede03a990e220fad95cd2614fb8af7edddc passed [CI](https://github.com/containerd/containerd/runs/1713224064).

katiewasnothere commented on diff/windows/windows.go:
```diff
@@ -191,3 +326,54 @@ func mountsToLayerAndParents(mounts []mount.Mount) (string, []string, error) {
+		return nil, errors.Wrapf(err, "Lower mount invalid")
+	}
+
+	if upperParentLayerPaths[0] != lowerLayer || len(upperParentLayerPaths) != len(lowerParentLayerPaths)+1 {
+		return nil, errors.Wrap(errdefs.ErrInvalidArgument, "windowsDiff cannot diff a layer against a layer other than its own parent")
+	}
```
Do we want to return a separate error for the second case in this if clause that indicates we have an unexpected number of upper layers?

TBBle replied:
Good point. The length check is a pre-condition to the following loop, and should return the same error, since what we actually care about is that `upperParentLayerPaths[1:] == lowerParentLayerPaths`.

katiewasnothere commented:
@TBBle do you have any plans to work on the LCOW version of `Compare` as well?

TBBle commented:
I have no specific plans for LCOW work. The motivating case (and initial test-case, before I noticed recently that there were a couple of integration tests disabled due to lack of this code) is BuildKit, who have been specifically disinterested in LCOW support, so it never go particularly high on my list.

I also don't currently know how LCOW layer storage works, as I've been particularly focussed on WCOW, (and specifically, if WSL2 or Windows 2004 have affected LCOW, I'm not going to be in a position to test RS5 personally) so all I've been looking for here is to not _break_ LCOW for some future person.

Of course, once WCOW is landed and working, idle curiosity might bring me back to LCOW. It's not clear to me how important LCOW is in the Windows Containers ecosystem these days, versus Linux-Docker-in-WSL2 for example. k8s mixed-OS Pods might make that more interesting, if they want to go that way.

katiewasnothere commented on .github/workflows/ci.yml:
```diff
@@ -206,9 +206,9 @@ jobs:
           mingw32-make.exe binaries
           bindir="$(pwd)"
-          SHIM_COMMIT=$(grep 'Microsoft/hcsshim ' go.mod | awk '{print $2}')
+          SHIM_COMMIT=$(grep -m 1 'Microsoft/hcsshim ' go.mod | sed -e 's#.*[ -]##')
```
Why this?

TBBle replied:
Two reasons:
* This command fails if the hcsshim being pointed to is a commit, not a tag, hence the change from `awk` to `sed`.
* Part of my testing for this stack was to [add a `replace` entry for hcsshim](https://github.com/containerd/containerd/pull/4419/commits/211176521cdba964c96bd7319f1eb28c216f385f) and hence `grep` finds multiple instances of `Microsoft/hcsshim`, hence the addition of `-m 1` to only take the first match.

See 260cb0b46eb497ee5ed70a4ff535ae064854acb4

TBBle replied:
Following up on this, a new hcsshim release v0.8.15 includes the necessary changes, and for some reason this code was not seeing that tag on CI in the `git fetch origin` call, so I've dropped this change for now.

Since I'm using `replaces` for testing elsewhere (I don't need the shim to be updated AFAIK), I suspect I'll only need the `-m 1` part when doing that, but this PR doesn't need this change.

katiewasnothere commented on diff/windows/windows.go:
```diff
@@ -191,3 +326,54 @@ func mountsToLayerAndParents(mounts []mount.Mount) (string, []string, error) {
+	return append([]string{upperLayer}, upperParentLayerPaths...), nil
+}
+
+func uniqueRef() string {
+	t := time.Now()
+	var b [3]byte
+	// Ignore read failures, just decreases uniqueness
+	rand.Read(b[:])
+	return fmt.Sprintf("%d-%s", t.UnixNano(), base64.URLEncoding.EncodeToString(b[:]))
+}
```
This looks like the same function as defined in diff/walking/differ.go maybe we should move it to a shared utils file?

TBBle replied:
I indeed just copied it from there. That's related to [what I wrote earlier](https://github.com/containerd/containerd/pull/4399#issuecomment-660283335): the `Compare` function it's called from is also basically the same as the one in the naive differ, and they should be refactored to share the common parts. I copied this one as-is because I wasn't sure we'd want to make this a public API and (at the time) I was unaware of things like the `internal/` package tree.

I think this function also appears in a third place in the codebase, too.

TBBle replied:
I've identified five copies of this function (including the new one I've added here), plus one semantic-equivalent:
* unpacker.go `uniquePart` (uses `t.Nanosecond()`)
* diff/walking/differ.go `uniqueRef` (uses `t.UnixNano()`) -- This is the one I copied, because `(s windowsDiff) Compare` itself started life as a copy of `(s *walkingDiff) Compare`.
* diff/windows/windows.go `uniqueRef` (uses `t.UnixNano()`) -- The new one I'm adding here.
* leases/id.go `WithRandomID`  (uses `t.Nanosecond()`)
* pkg/cri/util/id.go `GenerateID` (no timestamp, uses `hex.EncodeToString` instead of `base64.URLEncoding.EncodeToString`)
* rootfs/apply.go `uniquePart` (uses `t.Nanosecond()`)

I honestly don't see a good place to put a unified copy of this, structure-wise. I don't see a util or helper package, and there isn't an _/internal_ tree at this time to add module-local helpers to.

katiewasnothere commented:
> It's not clear to me how important LCOW is in the Windows Containers ecosystem these days, versus Linux-Docker-in-WSL2 for example. k8s mixed-OS Pods might make that more interesting, if they want to go that way.

I may be a bit biased there😋 but that's fine to not implement if there's no use case for it right now.

TBBle commented:
Oops, somehow this fell off my 'unread' list with unaddressed feedback. I'll try and get the comments resolved this week, so I can try and land this, and rebase-and-land #4415 early in the 1.6.0 containerd cycle, so hopefully I can get #4419 delivered for 1.6.0.

TBBle commented:
I've rebased to resolve some conflicts, and addressed one comment with a clear and simple resolution, but I don't at this time have resolutions for the other comments that I'm satisfied with, or at least more-satisfied with than "punting it to later".

TBBle commented:
Latest rebase hit #4924 in the WIndows Integration stage at the end of the first integration test suite run, which otherwise passed.

katiewasnothere commented:
@kevpar or @jterry75 could one of you take another look?

TBBle commented:
WIndows CI test-run failure was in `TestConvert` and appears to be caused by a temporary HTTP 500 error from mcr.microsoft.com, unrelated to my changes:
```
2021-03-13T00:40:11.9550601Z     log_hook.go:40: time="2021-03-13T00:40:11.952829500Z" level=debug msg="fetch response received" digest="sha256:c2a53c300c9474f1d971aed19c441933a295d19fbc5f45fa1f39d3a2edd0bcfc" mediatype=application/vnd.docker.container.image.v1+json response.header.content-length=26 response.header.date="Sat, 13 Mar 2021 00:40:11 GMT" response.header.x-azure-internalerror=0x830a1919 response.header.x-msedge-ref="Ref A: 80BE0FA83B6D44D38700D7479D7D590F Ref B: ASHEDGE1309 Ref C: 2021-03-13T00:40:11Z" response.status="500 Internal Server Error" size=393 url="https://mcr.microsoft.com/v2/windows/nanoserver/blobs/sha256:c2a53c300c9474f1d971aed19c441933a295d19fbc5f45fa1f39d3a2edd0bcfc"
2021-03-13T00:40:11.9649875Z     convert_test.go:50: failed to copy: httpReadSeeker: failed open: unexpected status code https://mcr.microsoft.com/v2/windows/nanoserver/blobs/sha256:c2a53c300c9474f1d971aed19c441933a295d19fbc5f45fa1f39d3a2edd0bcfc: 500 Internal Server Error
```

`TestExport` and `TestExportAndImport`, the newly-unskipped tests in this PR, passed.

jterry75 commented:
@TBBle - This looks good to me. Do you have any more work to do on this one?

TBBle commented:
Well, looks like I have a conflict to resolve, but apart from that, I reckon it's ready-to-go.

kzys commented:
@katiewasnothere @jterry75 Looks good to me! Can you folks take a look once again and merge the change if that looks good?

ambarve commented:
This looks good to me. @jterry75  can you please merge this?

TBBle commented:
Hmm. Rebasing showed a random bunch of HTTP-related failures. The Windows Integration run showed:
```
2021-04-05T04:02:48.2118915Z     convert_test.go:50: failed to copy: stream error: stream ID 9; HTTP_1_1_REQUIRED
2021-04-05T04:02:48.2120056Z --- FAIL: TestConvert (5.78s)
```
which is clearly unrelated to my changes here. The Linux suite are all failing on CRI tests, getting '403 forbidden' from GCR for some image or images.

jterry75 commented:
@TBBle - Should we just re-run the tests are are you going ot rebase on some master change that is clean?

TBBle commented:
I can't retrigger the tests AFAIK, but if you can, that'd be fine. I had a poke around, and the failures looked more like infrastructure issues than code issues, so I am not specifically waiting on a fix to rebase on-to.

I only rebased this time because I needed to rebase #4419 to verify a change, and figured I'd do the whole stack to keep my `git show-branches` output rational.

AkihiroSuda commented on diff/windows/windows.go:
```diff
@@ -158,7 +164,136 @@ func (s windowsDiff) Apply(ctx context.Context, desc ocispec.Descriptor, mounts
+	case ocispec.MediaTypeImageLayerGzip:
+		isCompressed = true
+	default:
+		return emptyDesc, errors.Wrapf(errdefs.ErrNotImplemented, "unsupported diff media type: %v", config.MediaType)
```
Can we support Docker media types?

TBBle replied:
Ummm. Weird. The [walking difffer](https://github.com/containerd/containerd/blob/dd14169771817b2674c22240ca63ba54bdd0f803/diff/walking/differ.go#L68-L79) doesn't list docker types here either.

Is it possible that a different layer has already folded the Docker media types into their OCI equivalents, i.e. are we passing through, e.g., [here](https://github.com/containerd/containerd/blob/a1138182d5d746351b3c321893001e33fd7fb4be/images/converter/default.go#L79-L93) or similar first, for all the code-paths that arrive here?

Otherwise, I think that might be a different change, that could be applied to both existing `Compare` implementations at once. The third differ, LCOW, doesn't have an implementation of `Compare` yet.

TBBle replied:
Following up on this, I did notice that `TestImport` covers [some Docker media types](https://github.com/containerd/containerd/blob/85041ffe58751b27e3fc068a1d605e058d12e4aa/integration/client/import_test.go#L143-L176), but `TestExportAndImport` and `TestExport` do not, as it appears that the Export format always produces OCI layers per the OCI image spec, and then adds in a Docker-compatible manifest.

In fact, I can't see any code-path except `ctr snapshot diff --media-type` that could get a media type setting into this API, which is a _huge_ hole in the test suite. We do expose `diff.WithMediaType` for use with `client.DiffService().Compare` and a `MediaType` field in `api/services/diff/v1.DiffRequest` for API users who might reasonably expect to be able to get Docker layer types out, since they were able to *import* Docker media types, but we never test those options, that I can see. >_<

TBBle replied:
I created #5317 so this isn't forgotten.

fuweid commented on diff/windows/windows.go:
```diff
@@ -191,3 +326,58 @@ func mountsToLayerAndParents(mounts []mount.Mount) (string, []string, error) {
 	return mnt.Source, parentLayerPaths, nil
 }
+
+// Ensure the two sets of mount-lists are actually a correct parent-and-child, or
```
nit: comment should start with `mountPairToLayerStack` :)

fuweid reviewed (commented):
lgtm on green

TBBle commented:
Gah. The rebase to current mainline branch passed: https://github.com/containerd/containerd/actions/runs/725556015
And then a comment-only fix failed in unrelated tests on Windows: https://github.com/containerd/containerd/actions/runs/725563341

```
2021-04-07T10:00:33.8681308Z === CONT  TestContainerKill
2021-04-07T10:00:33.8694170Z     container_test.go:620: failed to mount container storage: hcsshim::PrepareLayer - failed failed in Win32: The system cannot find the path specified. (0x3): unknown
2021-04-07T10:00:33.9335940Z --- FAIL: TestContainerKill (2.32s)
```
```
2021-04-07T10:01:33.7263019Z === CONT  TestContainerNoBinaryExists
2021-04-07T10:01:33.7264655Z     container_test.go:759: failed to create task failed to mount container storage: hcsshim::PrepareLayer - failed failed in Win32: The device is not ready. (0x15): unknown
2021-04-07T10:01:33.7552071Z --- FAIL: TestContainerNoBinaryExists (62.14s)
```

It's green-enough for me, since this PR doesn't actually change any of the on-disk layer data, and these look suspiciously like the HCS-was-busy-and-reported-success-too-early issue I speculated in #4419.

Notably, the two tests this PR enables, `TestExport` and `TestExportAndImport`, had passed before this point in the run
```
2021-04-07T09:59:53.0784565Z --- PASS: TestExport (5.86s)
2021-04-07T10:00:30.3192956Z --- PASS: TestExportAndImport (15.25s)
```

jterry75 commented:
@TBBle - This looks to be a longstanding bug with the containerd shim here: https://github.com/microsoft/hcsshim/blob/master/cmd/containerd-shim-runhcs-v1/delete.go#L29. The code properly handles the vm/container not found but I believe unintentionally re-uses the `err` defined when the span status is set here: https://github.com/microsoft/hcsshim/blob/master/cmd/containerd-shim-runhcs-v1/delete.go#L36. I believe this is truly unrelated to your change and will sign off.

Thanks!

jterry75 reviewed (approved):
LGTM

jterry75 commented:
@fuweid - I looked at the Windows error and agree this is unrelated to this change. Are you ok taking it with the transient error while @ambarve can get a fix out?

fuweid commented:
@TBBle would you mind to rebase? Your change might be not related to the error. Please rebase and retry it again. Thanks

jterry75 commented:
All green!

fuweid reviewed (approved):
LGTM
